### ISSUE TITLE
"enoent no such file or directory" - only when applying patches

### ISSUE THREAD

**Issue by getglad** [Snyk/snyk#22]
- `node -v`: 4.4.3
- `npm -v`: 3.8.6
- `snyk -v`: 1.14.0
- Command run: snyk wizard --dev

### Expected behaviour
Apply patches

### Actual behaviour
Get warning message "enoent no such file or directory" in UserName\AppData\Local\Temp\Temp_File_Name

### Steps to reproduce
- Using Windows 7
- Accept a patch

I'm not sure what else is helpful, but let me know and I can share.

**Comment by remy** [comment]
If you can re-run with `-d` as the argument - this gives full debug output. Note however you may want to remove any sensitive information, or equally email it to support@snyk.io and we can take a look offline.

**Comment by getglad** [comment]
Here you go:

```
Your environment has been set up for using Node.js 4.4.3 (x64) and npm.

U:\>cd C:/test/snyk-test

U:\>C:

C:\test\snyk-test>snyk wizard --dev -d
  snyk wizard +0ms { _: [ [Circular] ], dev: true, debug: true }
  snyk ~~~~ LIVE RUN ~~~~ +1s
  snyk checking for cli updates +4ms
  snyk add +707ms inline-auth false

Snyk's wizard will:

  * Enumerate your local dependencies and query Snyk's servers for vulnerabiliti
es
  * Guide you through fixing found vulnerabilities
  * Create a .snyk policy file to guide snyk commands such as `test` and `protec
t`
  * Remember your dependencies to alert you when new vulnerabilities are disclos
ed

  snyk add +10ms local true
  snyk policies found +5s [ 'C:\\test\\snyk-test' ]
  snyk add +2ms policies 1
  snyk add +1ms packageName snykTest
  snyk add +1ms packageVersion 0.0.0
  snyk add +1ms package snykTest@0.0.0
  snyk add +14s vulns-pre-policy 16...
  snyk add +5ms vulns 16es database...
Querying vulnerabilities database...
Tested 947 dependencies for known vulnerabilities, found 9 vulnerabilities, 16 v
ulnerable paths.
  snyk starting questions +4ms
  snyk getUpdatePrompts +1ms
  snyk + adding hawk@3.1.3 to upgrades +13ms
  snyk + adding request@2.68.0 to upgrades +1ms

  snyk final show npm:ws:20160104 when true +13ms
? 2 vulnerabilities introduced via gulp-protractor@2.1.0
- info: https://snyk.io/package/npm/gulp-protractor/2.1.0
  Remediation options Upgrade to gulp-protractor@2.3.0
  snyk no group answer: show npm:ws:20160104 when false +3s
  snyk no group answer: show npm:hawk:20160119 when false +1ms

  snyk final show npm:request:20160119 when true +3ms
? 2 vulnerabilities introduced via gulp-sass@2.0.4
- info: https://snyk.io/package/npm/gulp-sass/2.0.4
  Remediation options Re-install gulp-sass@2.0.4
  snyk no group answer: show npm:request:20160119 when false +2s
  snyk no group answer: show npm:request:20160119 when false +2ms

  snyk final show npm:uglify-js:20151024 when true +1ms
? Low severity vulnerability found in uglify-js@2.5.0, introduced via gulp-ugli
fy@1.4.2
- desc: Regular Expression Denial of Service
- info: https://snyk.io/vuln/npm:uglify-js:20151024
- from: gulp-uglify@1.4.2 > uglify-js@2.5.0
  Remediation options Upgrade to gulp-uglify@1.5.1
  snyk getPatchPrompts +2s
  snyk found alternative location for uglify-js@2.2.5 (<= 2.4.23 >=2.2.0 by npm:
uglify-js:20150824) in vuln npm:uglify-js:20151024 +11ms

  snyk final show npm:request:20160119 when true +5ms
? 3 vulnerabilities introduced in request
- info: https://snyk.io/package/npm/request/2.67.0
  Remediation options Patch
  snyk no group answer: show npm:request:20160119 when false +927ms
  snyk no group answer: show npm:request:20160119 when false +1ms
  snyk no group answer: show npm:request:20160119 when false +1ms

  snyk final show npm:uglify-js:20151024 when true +3ms
? 2 vulnerabilities introduced in uglify-js
- info: https://snyk.io/package/npm/uglify-js/2.2.5
  Remediation options Patch
  snyk no group answer: show npm:uglify-js:20151024 when false +644ms
  snyk no group answer: show npm:uglify-js:20150824 when false +0ms
  snyk getIgnorePrompts +2ms

  snyk final show npm:connect:20130701 when true +10ms
? Medium severity vulnerability found in connect@1.9.2, introduced via browser-
sync@2.9.12
- desc: methodOverride Middleware Reflected Cross-Site Scripting
- info: https://snyk.io/vuln/npm:connect:20130701
- from: browser-sync@2.9.12 > browser-sync-ui@0.5.19 > weinre@2.0.0-pre-I0Z7U9O
V > express@2.5.11 > connect@1.9.2
  Remediation options Skip

  snyk final show npm:express:20140912 when true +2s
? Medium severity vulnerability found in express@2.5.11, introduced via browser
-sync@2.9.12
- desc: Cross-Site Scripting
- info: https://snyk.io/vuln/npm:express:20140912
- from: browser-sync@2.9.12 > browser-sync-ui@0.5.19 > weinre@2.0.0-pre-I0Z7U9O
V > express@2.5.11
  Remediation options Skip

  snyk final show npm:qs:20140806 when true +621ms
? Medium severity vulnerability found in qs@0.4.2, introduced via browser-sync@
2.9.12
- desc: Denial of Service (Memory Exhaustion)
- info: https://snyk.io/vuln/npm:qs:20140806
- from: browser-sync@2.9.12 > browser-sync-ui@0.5.19 > weinre@2.0.0-pre-I0Z7U9O
V > express@2.5.11 > qs@0.4.2
  Remediation options Skip

  snyk final show npm:qs:20140806-1 when true +541ms
? Medium severity vulnerability found in qs@0.4.2, introduced via browser-sync@
2.9.12
- desc: Denial of Service (Event Loop Blocking)
- info: https://snyk.io/vuln/npm:qs:20140806-1
- from: browser-sync@2.9.12 > browser-sync-ui@0.5.19 > weinre@2.0.0-pre-I0Z7U9O
V > express@2.5.11 > qs@0.4.2
  Remediation options Skip
? Add `snyk test` to package.json file to fail test on newly disclosed vulnerab
ilities? Yes
? Add `snyk protect` as a package.json installation hook to apply chosen patche
s on install? Yes
  snyk add +3s answers [ { vulnId: 'npm:ws:20160104', choice: 'update' },
  { vulnId: 'npm:request:20160119', choice: 'update' },
  { vulnId: 'npm:uglify-js:20151024', choice: 'update' },
  { vulnId: 'npm:request:20160119', choice: 'patch' },
  { vulnId: 'npm:uglify-js:20151024', choice: 'patch' },
  { vulnId: 'npm:connect:20130701', choice: 'skip' },
  { vulnId: 'npm:express:20140912', choice: 'skip' },
  { vulnId: 'npm:qs:20140806', choice: 'skip' },
  { vulnId: 'npm:qs:20140806-1', choice: 'skip' } ]
  snyk additional answers required: 2 +3ms { affected: { name: 'request', versio
n: '2.67.0', full: 'request' },
  main: true,
  id: 'npm:request:20160119-5',
  count: 3,
  upgrades:
   [ [ 'snykTest@0.0.0', 'phantomjs@1.9.20', 'request@2.67.0' ],
     [ 'snykTest@0.0.0',
       'gulp-protractor@2.1.0',
       'protractor@3.0.0',
       'request@2.57.0' ],
     [ 'snykTest@0.0.0',
       'browser-sync@2.9.12',
       'localtunnel@1.8.1',
       'request@2.65.0' ] ],
  patch: true }
  snyk additional answers required: 1 +3ms { affected: { name: 'uglify-js', vers
ion: '2.2.5', full: 'uglify-js' },
  main: true,
  id: 'npm:uglify-js:20151024-8',
  count: 2,
  upgrades:
   [ [ 'snykTest@0.0.0',
       'jade@1.11.0',
       'transformers@2.1.0',
       'uglify-js@2.2.5' ] ],
  patch: true,
  includes: [ 'npm:uglify-js:20150824' ] }
  snyk { ignore: [], update: [ { title: 'Remote Memory Exposure', creationTime:
'2016-01-05T12:38:01.749Z', modificationTime: '2016-01-05T12:38:01.749Z', public
ationTime: '2016-01-04T19:34:19.734Z', semver: [Object], CVSSv3: '', severity: '
high', identifiers: [Object], patches: [Object], moduleName: 'ws', id: 'npm:ws:2
0160104', from: [Object], upgradePath: [Object], version: '0.8.1', name: 'ws', i
sUpgradable: true, isPatchable: true, __filename: 'C:\\test\\snyk-test\\node_mod
ules\\selenium-webdriver\\node_modules\\ws\\package.json', shrinkwrap: undefined
, bundled: undefined, grouped: [Object] }, { title: 'Remote Memory Exposure', cr
eationTime: '2016-03-22T12:00:05.158Z', modificationTime: '2016-03-22T12:00:05.1
58Z', publicationTime: '2016-01-19T04:57:05.158Z', semver: [Object], CVSSv3: '',
 severity: 'medium', identifiers: [Object], patches: [Object], moduleName: 'requ
est', id: 'npm:request:20160119', from: [Object], upgradePath: [Object], version
: '2.65.0', name: 'request', isUpgradable: true, isPatchable: true, __filename:
'C:\\test\\snyk-test\\node_modules\\request\\package.json', shrinkwrap: undefine
d, bundled: undefined, grouped: [Object] }, { title: 'Regular Expression Denial
of Service', creationTime: '2015-11-06T02:09:36.186Z', modificationTime: '2015-1
1-06T02:09:36.186Z', publicationTime: '2015-10-24T21:01:32.155Z', semver: [Objec
t], CVSSv3: '', severity: 'low', identifiers: [Object], patches: [Object], modul
eName: 'uglify-js', id: 'npm:uglify-js:20151024', from: [Object], upgradePath: [
Object], version: '2.5.0', name: 'uglify-js', isUpgradable: true, isPatchable: t
rue, __filename: 'C:\\test\\snyk-test\\node_modules\\gulp-uglify\\node_modules\\
uglify-js\\package.json', shrinkwrap: undefined, bundled: undefined } ], patch:
[ { title: 'Remote Memory Exposure', creationTime: '2016-03-22T12:00:05.158Z', m
odificationTime: '2016-03-22T12:00:05.158Z', publicationTime: '2016-01-19T04:57:
05.158Z', semver: [Object], CVSSv3: '', severity: 'medium', identifiers: [Object
], patches: [Object], moduleName: 'request', id: 'npm:request:20160119', from: [
Object], upgradePath: [Object], version: '2.67.0', name: 'request', isUpgradable
: false, isPatchable: true, __filename: 'C:\\test\\snyk-test\\node_modules\\phan
tomjs\\node_modules\\request\\package.json', shrinkwrap: undefined, bundled: und
efined, grouped: [Object] }, { title: 'Remote Memory Exposure', creationTime: '2
016-03-22T12:00:05.158Z', modificationTime: '2016-03-22T12:00:05.158Z', publicat
ionTime: '2016-01-19T04:57:05.158Z', semver: [Object], CVSSv3: '', severity: 'me
dium', identifiers: [Object], patches: [Object], moduleName: 'request', id: 'npm
:request:20160119', from: [Object], upgradePath: [Object], version: '2.67.0', na
me: 'request', isUpgradable: false, isPatchable: true, __filename: 'C:\\test\\sn
yk-test\\node_modules\\phantomjs\\node_modules\\request\\package.json', shrinkwr
ap: undefined, bundled: undefined, grouped: [Object] }, { title: 'Remote Memory
Exposure', creationTime: '2016-03-22T12:00:05.158Z', modificationTime: '2016-03-
22T12:00:05.158Z', publicationTime: '2016-01-19T04:57:05.158Z', semver: [Object]
, CVSSv3: '', severity: 'medium', identifiers: [Object], patches: [Object], modu
leName: 'request', id: 'npm:request:20160119', from: [Object], upgradePath: [Obj
ect], version: '2.67.0', name: 'request', isUpgradable: false, isPatchable: true
, __filename: 'C:\\test\\snyk-test\\node_modules\\phantomjs\\node_modules\\reque
st\\package.json', shrinkwrap: undefined, bundled: undefined, grouped: [Object]
}, { title: 'Regular Expression Denial of Service', creationTime: '2015-11-06T02
:09:36.186Z', modificationTime: '2015-11-06T02:09:36.186Z', publicationTime: '20
15-10-24T21:01:32.155Z', semver: [Object], CVSSv3: '', severity: 'low', identifi
ers: [Object], patches: [Object], moduleName: 'uglify-js', id: 'npm:uglify-js:20
151024', from: [Object], upgradePath: [Object], version: '2.2.5', name: 'uglify-
js', isUpgradable: false, isPatchable: true, __filename: 'C:\\test\\snyk-test\\n
ode_modules\\transformers\\node_modules\\uglify-js\\package.json', shrinkwrap: u
ndefined, bundled: undefined, grouped: [Object] } ], skip: [] } +1ms
  snyk to upgrade +41ms [ 'gulp-protractor@2.3.0',
  'gulp-sass@2.0.4',
  'gulp-uglify@1.5.1' ]
  snyk C:\test\snyk-test$ npm --save uninstall gulp-protractor gulp-sass gulp-ug
lify +1ms
  snyk patching 2 +9ms


┌────────────────────────────────────────────┐
│ Update available: 1.14.1 (current: 1.14.0) │
│ Run npm install -g snyk to update.         │
└────────────────────────────────────────────┘

stream.js:74
      throw er; // Unhandled stream error in pipe.
      ^

Error: ENOENT: no such file or directory, open 'C:\Users\...\AppData\Local\
Temp\2122a3d9-b968-483f-aeca-2224f0952335.npm:request:20160119.snyk-patch'
    at Error (native)

C:\test\snyk-test>
```

**Comment by remy** [comment]
Can you try something for me (I don't have a windows machine to hand this second, but it might be a quick way of testing an issue).

Can you try to create an empty file called `2224f0952335.npm:request:20160119.snyk-patch` - and tell me if that works (specific if the `:` values remain in the filename)?

Thanks,

**Comment by getglad** [comment]
Windows will not let me create a file with a `:` in it - neither manually or through the command prompt.

**Comment by remy** [comment]
Perfect, that's what I was hoping for. I'll get a fix early next week and
I'll ping you here.

**Comment by remy** [comment]
There's a PR waiting to be merged (though it'll be tomorrow AM) and I'll push a release out on the same day. I'll ping you as soon as it's out and hopefully you'll be able to give it a go on Windows 7.

**Comment by getglad** [comment]
Ready and willing :)

**Comment by remy** [comment]
Can you give the latest snyk a test? 1.14.2 includes the fix for Windows 7 and removes the use of `:` in filenames: https://github.com/Snyk/snyk/releases/tag/v1.14.2

**Comment by getglad** [comment]
Hate to post this but...

```
...
? Add `snyk test` to package.json file to fail test on newly disclosed vulnerab
ilities? Yes
? Add `snyk protect` as a package.json installation hook to apply chosen patche
s on install? Yes
Running `npm update`...
Applying patches...
ENOENT: no such file or directory, open 'C:\test\snyk-test\node_modules\request\
.snyk-npm:request:20160119.flag'

C:\test\snyk-test>snyk -v
1.14.2
```

Looks like 1 file name didn't get updated

**Comment by remy** [comment]
Damn. Thanks for testing, I'll take a run at it again today. Thought I had them all, clearly not!

**Comment by getglad** [comment]
Is there anything I can help out with on this issue? Happy to give it go if you want to tell me where to look/start.

**Comment by harriha** [comment]
Reporting back, just in case: this seems to be still there with the version 1.15.0 which was released a moment ago (no mention about fixes in release notes so this is as expected). I'm on Windows 10.

After doing some further testing (with 1.15.0), I can see the following:

- I seem to see this error message only if answering 'No' to `? Do you want to revisit your existing policy [y] or only update it [N]?`. If answering 'Yes', there's no error but the patching doesn't work either even though the CLI flow seems fine. After finishing it, I see nothing in `.snyk` file's `patch` prop. Pleas see attached flow if that helps. 
- If observing the actual module snyk tries to patch, seems that it gets done, there's original source file as well as the `.orig` file present. However, `snyk test` keeps nagging about the problem (assumably because that problematic meta file is missing).


```
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$ snyk test                                                                                             
✗ High severity vulnerability found on minimatch@2.0.10                                                 
- desc: Regular Expression Denial of Service                                                            
- info: https://snyk.io/vuln/npm:minimatch:20160620                                                     
- from: my-repository@0.3.0 > nodemon@1.9.2 > chokidar@1.5.2 > readdirp@2.0.0 > minimatch@2.0.10   
No direct dependency upgrade can address this issue.                                                    
Run `snyk wizard` to explore remediation options.                                                       
                                                                                                        
Tested 470 dependencies for known vulnerabilities, found 1 vulnerability, 1 vulnerable path.            
                                                                                                        
Run `snyk wizard` to address these issues.                                                              
                                                                                                        
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$ snyk wizard                                                                                           
                                                                                                        
Snyk's wizard will:                                                                                     
                                                                                                        
  * Enumerate your local dependencies and query Snyk's servers for vulnerabilities                      
  * Guide you through fixing found vulnerabilities                                                      
  * Create a .snyk policy file to guide snyk commands such as `test` and `protect`                      
  * Remember your dependencies to alert you when new vulnerabilities are disclosed                      
                                                                                                        
? Do you want to revisit your existing policy [y] or only update it [N]? Yes                            
Querying vulnerabilities database...                                                                    
Tested 470 dependencies for known vulnerabilities, found 1 vulnerabilities, 2 vulnerable paths.         
                                                                                                        
? High severity vuln found in minimatch@2.0.10, introduced via nodemon@1.9.2                            
- desc: Regular Expression Denial of Service                                                            
- info: https://snyk.io/vuln/npm:minimatch:20160620                                                     
- from: nodemon@1.9.2 > chokidar@1.5.2 > readdirp@2.0.0 > minimatch@2.0.10                              
  Remediation options Patch                                                                             
                                                                                                        
? High severity vuln found in minimatch@0.3.0, introduced via snyk@1.15.0                               
- desc: Regular Expression Denial of Service                                                            
- info: https://snyk.io/vuln/npm:minimatch:20160620                                                     
- from: snyk@1.15.0 > recursive-readdir@1.3.0 > minimatch@0.3.0                                         
  Remediation options Ignore                                                                            
? [audit] Reason for ignoring vulnerability? None given                                                 
Saving .snyk policy file...                                                                             
Remembering current dependencies for future notifications...                                            
                                                                                                        
Your .snyk policy file has been successfully updated.                                                   
To review your policy, run `snyk policy`.                                                               
                                                                                                        
You can see a snapshot of your dependencies here:                                                       
https://snyk.io/org/<CENSORED>                                
                                                                                                        
We'll notify you when relevant new vulnerabilities are disclosed.                                       
                                                                                                        
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$ snyk test                                                                                             
✗ High severity vulnerability found on minimatch@2.0.10                                                 
- desc: Regular Expression Denial of Service                                                            
- info: https://snyk.io/vuln/npm:minimatch:20160620                                                     
- from: my-repository@0.3.0 > nodemon@1.9.2 > chokidar@1.5.2 > readdirp@2.0.0 > minimatch@2.0.10   
No direct dependency upgrade can address this issue.                                                    
Run `snyk wizard` to explore remediation options.                                                       
                                                                                                        
Tested 470 dependencies for known vulnerabilities, found 1 vulnerability, 1 vulnerable path.            
                                                                                                        
Run `snyk wizard` to address these issues.                                                              
                                                                                                        
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$ snyk protect                                                                                          
Successfully applied Snyk patches                                                                       
                                                                                                        
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$ snyk test                                                                                             
✗ High severity vulnerability found on minimatch@2.0.10                                                 
- desc: Regular Expression Denial of Service                                                            
- info: https://snyk.io/vuln/npm:minimatch:20160620                                                     
- from: my-repository@0.3.0 > nodemon@1.9.2 > chokidar@1.5.2 > readdirp@2.0.0 > minimatch@2.0.10   
No direct dependency upgrade can address this issue.                                                    
Run `snyk wizard` to explore remediation options.                                                       
                                                                                                        
Tested 470 dependencies for known vulnerabilities, found 1 vulnerability, 1 vulnerable path.            
                                                                                                        
Run `snyk wizard` to address these issues.                                                              
                                                                                                        
harri@FL671 MINGW64 ~/repos/my-repository (master)                                                 
$
```

**Comment by getglad** [comment]
I'm kinda stuck on testing at the moment - our network guys have reworked all inbound SSL traffic and caused package managers to get angry - https://github.com/Snyk/snyk/issues/15#issuecomment-227740813

**Comment by smockle** [comment]
Hi, I opened a [PR](https://github.com/Snyk/snyk/pull/30) that I think resolves this. There are a few caveats listed in the PR, but `npm i --save git://github.com/smockle/snykier.git#v1.16.0` ought to let anyone interested test the fix.

**Comment by remy** [comment]
Should be fixed in 1.5.1 - https://github.com/Snyk/snyk/releases

Thanks all.
